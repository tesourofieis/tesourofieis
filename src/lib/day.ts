import { getDate, getDay, getMonth } from "date-fns";
import {
  COMMUNE_C_10A,
  COMMUNE_C_10B,
  COMMUNE_C_10C,
  COMMUNE_C_10PASC,
  COMMUNE_C_10T,
  PATTERN_ADVENT,
  PATTERN_EASTER,
  PATTERN_EPIPHANY,
  PATTERN_LENT,
  PATTERN_POST_EPIPHANY_SUNDAY,
  PATTERN_PRE_LENTEN,
  TABLE_OF_PRECEDENCE,
  TEMPORA_RANK_MAP,
  TITLES,
  TYPE_TEMPORA,
} from "./constants.ts";

// """ Class used to keep `Observance` objects for particular days of Missal.
//
// It contains three lists: `tempora`, `celebration` and `commemoration`.
// On Missal's creation the lists are filled in so that `tempora` always contains `Observance` representing
// given variable day, `celebration` contains an `Observance`s to be celebrated in this day and
// `commemoration` contains zero or more `Observance`s that should be commemorated with the main celebration.
// """
class Day {
  date: string;
  tempora: Observance[] = [];
  celebration: Observance[] = [];
  commemoration: Observance[] = [];

  constructor(date: string) {
    this.date = date;
  }

  get all(): Observance[] {
    return this.tempora.concat(this.celebration).concat(this.commemoration);
  }

  getTemporaName() {
    return this.tempora[0]?.title;
  }

  serialize() {
    const serialized = {};
    const containers = ["tempora", "celebration", "commemoration"];

    for (const container of containers) {
      serialized[container] = this[container].map((item: Day) =>
        item.serialize(),
      );
    }

    return serialized;
  }
}

class Observance {
  date: string;
  flexibility: string;
  name: string;
  rank: number;
  colors: string[];
  id: string;
  link: string;
  title: string;
  priority: number | null;
  weekday: number | null;

  // """ Build an Observance out of identifier and calendar date
  //
  // :param observance_id: observance's identifier in format
  //                <flexibility>:<identifier>:<rank>
  // :type observance_id: string
  // :param date_: specific date in which the observance is supposed
  //             to be placed. For some Sancti days its rank (class)
  //             depends on which calendar day they occur.
  // :type date_: `date ` object
  // """
  constructor(observanceId: string, date_: string) {
    const [flexibility, name, rank, color] = observanceId.split(":");
    this.date = date_;
    this.flexibility = flexibility;
    this.name = name;
    this.rank = this.calcRank(observanceId, Number.parseInt(rank, 10));
    this.colors = Array.from(color);
    this.id = `${this.flexibility}:${this.name}:${this.rank}:${color}`;
    this.link = `missal/${this.getTempora(this.name)}/${this.name}`;
    this.title = TITLES[observanceId];

    // Determine the weekday attribute based on the type of observance
    if (
      flexibility === TYPE_TEMPORA &&
      ![
        COMMUNE_C_10A,
        COMMUNE_C_10B,
        COMMUNE_C_10C,
        COMMUNE_C_10PASC,
        COMMUNE_C_10T,
      ].includes(observanceId)
    ) {
      if (name.replace(/^.*-(\d+).*$/, "$1") === "10-Dur") {
        this.weekday = 0;
      }
      this.weekday = Number.parseInt(name.replace(/^.*-(\d+).*$/, "$1"), 10);
    } else {
      this.weekday = getDay(this.date);
    }

    this.priority = this.calcPriority();
  }

  private getTempora(name: string): string {
    if (this.flexibility === "tempora") {
      if (name.includes("nat")) {
        return "natal";
      }
      if (name.includes("adv")) {
        return "advento";
      }
      if (name.includes("pasc")) {
        return "pascoa";
      }
      if (name.includes("epi")) {
        return "epifania";
      }
      if (name.includes("quadp")) {
        return "pre-quaresma";
      }
      if (name.includes("quad")) {
        return "quaresma";
      }
      if (name.includes("pent")) {
        return "pentecostes";
      }
    }

    if (this.flexibility === "santos") {
      console.log(`Matched Santos: ${name}`);
      return "santos";
    }

    if (this.flexibility === "commune") {
      return "comum";
    }

    if (this.flexibility === "votivas") {
      return "votivas";
    }

    console.log(`Checking name: ${name}, flexibility: ${this.flexibility}`);
    throw Error(`No directory for ${name}, flexibility , ${this.flexibility}`);
  }

  private calcRank(observanceId: string, originalRank: number): number {
    for (const caseItem of TEMPORA_RANK_MAP) {
      if (
        getMonth(this.date) === caseItem.month &&
        getDate(this.date) === caseItem.day &&
        new RegExp(caseItem.pattern).test(observanceId)
      ) {
        return caseItem.rank;
      }
    }
    return originalRank;
  }

  private calcPriority() {
    for (let priority = 0; priority < TABLE_OF_PRECEDENCE.length; priority++) {
      if (this.id.match(TABLE_OF_PRECEDENCE[priority])) {
        return priority;
      }
    }
  }

  serialize() {
    return {
      id: this.id,
      link: this.link,
      rank: this.rank,
      title: this.title,
      colors: this.colors,
    };
  }
}

export { Day, Observance };
