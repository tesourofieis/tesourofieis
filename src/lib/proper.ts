import {
  COMMEMORATED_ORATIO,
  COMMEMORATED_POSTCOMMUNIO,
  COMMEMORATED_SECRETA,
  COMMEMORATIONS,
  COMMEMORATIONS_PT,
  GRADUALE,
  GRADUALE_PASCHAL,
  NORMAL_SECTIONS,
  ORATIO,
  POSTCOMMUNIO,
  SECRETA,
  TRACTUS,
} from "./constants.ts";

export class ProperConfig {
  preface: string | null = null;
  interReadingsSection?: string | null = null;
  stripAlleluia = false;

  constructor(
    preface: string | null = null,
    stripAlleluia = false,
    interReadingsSection: string | null = null,
  ) {
    // interReadingsSection == null - show all sections defined in the source
    if (interReadingsSection !== null) {
      const validSections = [null, GRADUALE, TRACTUS, GRADUALE_PASCHAL];
      if (!validSections.includes(interReadingsSection)) {
        console.error(`Invalid interReadingsSection: ${interReadingsSection}`);
      }
    }
    this.preface = preface;
    this.interReadingsSection = interReadingsSection;
    this.stripAlleluia = stripAlleluia;
  }
}

export class Section {
  id = "";
  body: string[] = [];
  subSections: Section[] = [];

  constructor(id: string, body: string[] = []) {
    this.id = id;
    this.body = body || [];
  }

  getBody(): string[] {
    return this.body;
  }

  extendBody(bodyPart: string[]): void {
    this.body = [...this.body, ...bodyPart];
  }

  addSubsection(subsection: Section): void {
    this.subSections.push(subsection);
  }

  serialize() {
    return {
      id: this.id,
      body: this.body,
      subSections: this.subSections.map((i) => ({
        id: i.id,
        body: i.body,
      })),
    };
  }
}

export class ParsedSource {
  _container: Record<string, Section> = {};

  constructor() {
    this._container = {};
  }

  getSection(sectionId: string): Section | null {
    return this._container[sectionId] || null;
  }

  getSubSection(sectionId: string, subSectionId: string): Section | null {
    return (
      this._container[sectionId].subSections.find(
        (i) => i.id === subSectionId,
      ) || null
    );
  }

  setSection(sectionName: string, section: Section): void {
    this._container[sectionName] = section;
  }

  popSection(sectionId: string): Section | null {
    try {
      const body = this._container[sectionId];
      delete this._container[sectionId];
      return body;
    } catch (error) {
      return null;
    }
  }

  keys(): string[] {
    return Object.keys(this._container);
  }

  values(): Section[] {
    return Object.values(this._container);
  }

  items(): [string, Section][] {
    return Object.entries(this._container);
  }

  merge(proper: ParsedSource): void {
    for (const [k, v] of proper.items()) {
      if (!(k in this._container)) {
        this._container[k] = v;
      }
    }
  }
}

export class Proper extends ParsedSource {
  title: string | null = null;
  id: string;
  colors: string[];
  rank: number | null = null;
  commemorationsNamesTranslations:
    | typeof COMMEMORATIONS_PT
    | typeof COMMEMORATIONS;

  constructor(id: string, parsedSource: ParsedSource | null = null) {
    super();
    this.id = id;
    try {
      const [, , rank, color] = id.split(":");
      this.rank = parseInt(rank, 10);
      this.colors = color.split("");
    } catch (error) {
      throw new Error(`Proper ${id} not found`);
    }
    if (parsedSource !== null) {
      this._container = { ...parsedSource._container };
    }
  }

  serialize() {
    const list_ = this.values()
      .map((v) => v?.serialize())
      .sort((a, b) => {
        const indexA = NORMAL_SECTIONS.indexOf(a.id);
        const indexB = NORMAL_SECTIONS.indexOf(b.id);
        return indexA - indexB;
      });

    return list_;
  }

  getRule(ruleName: string): string | null {
    const rules: Record<string, string | null> = {
      preface: null,
      vide: null,
    };

    const rulesSrc: string[] = [];
    for (const sectionName of ["Rank", "Rule"]) {
      const section = this.getSection(sectionName);
      if (section !== null) {
        for (const line of section.getBody()) {
          rulesSrc.push(...line.split(";").map((i) => i.trim()));
        }
      }
    }

    if (rulesSrc.length > 0) {
      const preface = rulesSrc
        .filter((i) => i.startsWith("Prefatio") && i.includes("="))
        .map((i) => i.split("=")[1]);
      if (preface.length > 0) {
        rules.preface = preface[preface.length - 1];
      }

      const vide = rulesSrc
        .filter((i) => i.startsWith("vide ") || i.startsWith("ex "))
        .map((i) => i.split(" ").slice(-1)[0].split(";")[0]);
      if (vide.length > 0) {
        rules.vide = vide[0];
      }
    }

    return rules[ruleName] ?? null;
  }

  addCommemorations(commemorations: Proper[]) {
    for (const commemoration of commemorations) {
      const sections = [
        {
          commemorated_section_name: COMMEMORATED_ORATIO,
          source_section_name: ORATIO,
        },
        {
          commemorated_section_name: COMMEMORATED_SECRETA,
          source_section_name: SECRETA,
        },
        {
          commemorated_section_name: COMMEMORATED_POSTCOMMUNIO,
          source_section_name: POSTCOMMUNIO,
        },
      ];
      for (const {
        commemorated_section_name,
        source_section_name,
      } of sections) {
        const commemoratedSection =
          commemoration.getSection(source_section_name);
        commemoratedSection.body.unshift(commemoration.title);
        commemoratedSection.id = commemorated_section_name;
        this.setSection(commemorated_section_name, commemoratedSection);
      }
    }
  }
}
