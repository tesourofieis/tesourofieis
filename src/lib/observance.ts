import { getDate, getDay, getMonth } from "date-fns";
import {
  COMMUNE_C_10A,
  COMMUNE_C_10B,
  COMMUNE_C_10C,
  COMMUNE_C_10PASC,
  COMMUNE_C_10T,
  LANGUAGE,
  PATTERN_POST_EPIPHANY_SUNDAY,
  TABLE_OF_PRECEDENCE,
  TEMPORA_PENT23_0,
  TEMPORA_RANK_MAP,
  TYPE_TEMPORA,
  WEEKDAY_MAPPING,
  getTranslation,
} from "./constants.ts";
import { Proper, ProperConfig } from "./proper.ts";
import { ProperParser } from "./properParser.ts";

export class Observance {
  date: string;
  flexibility: string;
  name: string;
  rank: number;
  colors: string[];
  id: string;
  title: string;
  priority: number | null;
  weekday: number | null;

  // """ Build an Observance out of identifier and calendar date
  //
  // :param observance_id: observance's identifier in format
  //                <flexibility>:<identifier>:<rank>
  // :type observance_id: string
  // :param date_: specific date in which the observance is supposed
  //             to be placed. For some Sancti days its rank (class)
  //             depends on which calendar day they occur.
  // :type date_: `date ` object
  // """
  constructor(observanceId: string, date_: string) {
    const [flexibility, name, rank, color] = observanceId.split(":");
    this.date = date_;
    this.flexibility = flexibility;
    this.name = name;
    this.rank = this.calcRank(observanceId, parseInt(rank, 10));
    this.colors = Array.from(color);
    this.id = `${this.flexibility}:${this.name}:${this.rank}:${color}`;
    this.title = getTranslation(LANGUAGE).TITLES[observanceId];

    // Determine the weekday attribute based on the type of observance
    if (
      flexibility === TYPE_TEMPORA &&
      ![
        COMMUNE_C_10A,
        COMMUNE_C_10B,
        COMMUNE_C_10C,
        COMMUNE_C_10PASC,
        COMMUNE_C_10T,
      ].includes(observanceId)
    ) {
      this.weekday =
        WEEKDAY_MAPPING[parseInt(name.replace(/^.*-(\d+).*$/, "$1"), 10)];
    } else {
      this.weekday = getDay(this.date);
    }

    this.priority = this.calcPriority();
  }

  getProper(config?: ProperConfig) {
    const date = this.date;
    const proper = new ProperParser(this.id, config).parse();
    if (this.id.match(PATTERN_POST_EPIPHANY_SUNDAY) && getMonth(date) >= 9) {
      this.adjustSundayShiftedFromPostEpiphany(proper);
    }
    return proper;
  }

  hasProper(): boolean {
    return new ProperParser(this.id).properExists();
  }

  private calcRank(observanceId: string, originalRank: number): number {
    for (const caseItem of TEMPORA_RANK_MAP) {
      if (
        getMonth(this.date) === caseItem.month &&
        getDate(this.date) === caseItem.day &&
        new RegExp(caseItem.pattern).test(observanceId)
      ) {
        return caseItem.rank;
      }
    }
    return originalRank;
  }

  private calcPriority() {
    for (let priority = 0; priority < TABLE_OF_PRECEDENCE.length; priority++) {
      if (this.id.match(TABLE_OF_PRECEDENCE[priority])) {
        return priority;
      }
    }
  }

  private adjustSundayShiftedFromPostEpiphany(propers: Proper[]): void {
    const properSunday23PostPentecost = new ProperParser(
      TEMPORA_PENT23_0,
    ).parse();
    for (let i = 0; i < propers.length; i++) {
      const proper = propers[i];
      for (const section of [
        "INTROIT",
        "GRADUALE",
        "OFFERTORIUM",
        "COMMUNIO",
      ]) {
        proper.setSection(
          section,
          properSunday23PostPentecost[i][0]?.getSection(section),
        );
      }
    }
  }

  serialize() {
    return {
      id: this.id,
      rank: this.rank,
      title: this.title,
      colors: this.colors,
    };
  }
}
