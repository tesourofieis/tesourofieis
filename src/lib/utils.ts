import { CUSTOM_PREFACES, PROPERS_DIR } from "./constants.ts";
import path from "path";
import fs from "fs";
import { Observance } from "./observance.ts";
import { Proper } from "./proper.ts";
import { Day } from "./day.ts";
import type { Calendar } from "./calendar.ts";
import type { UTCDate } from "@date-fns/utc";
import { format } from "date-fns";

type Pattern = string | RegExp;

function match(
  observances: string | Observance | (string | Observance)[],
  patterns: string | Pattern | (string | Pattern)[],
): Observance | undefined {
  if (!Array.isArray(observances)) {
    observances = [observances];
  }

  if (!Array.isArray(patterns)) {
    patterns = [patterns];
  }

  for (const observance of observances) {
    const observanceId =
      typeof observance === "string" ? observance : observance.id;

    for (const pattern of patterns) {
      if (
        typeof pattern === "string" &&
        new RegExp(pattern).test(observanceId)
      ) {
        return observance as Observance;
      }
      if (pattern instanceof RegExp && pattern.test(observanceId)) {
        return observance as Observance;
      }
    }
  }
}

function yyyyMMDD(date: Date | UTCDate) {
  return format(date, "yyyy-MM-dd");
}

function getCustomPreface(
  celebration: Observance,
  tempora?: Observance,
): string | null {
  for (const { id, preface } of CUSTOM_PREFACES) {
    if (
      new RegExp(id).test(celebration.id) ||
      (tempora && celebration.rank > 1 && new RegExp(id).test(tempora.id))
    ) {
      return preface;
    }
  }
  return null;
}

function getPregeneratedProper(lang: string, properId: string) {
  const filePath = path.join(PROPERS_DIR, lang, `${properId}.json`);

  if (fs.existsSync(filePath)) {
    const fileContent = fs.readFileSync(filePath, "utf-8");
    return JSON.parse(fileContent);
  }

  return null;
}

export interface ProperDay {
  info: {
    id: string;
    title: string;
    tempora: string;
    rank: number;
    colors: string[];
    date: string;
    day: Day;
  };
  sections: Section[];
}

interface Section {
  id: string;
  body: {
    latin: string[];
    vernacular: string[];
  };
}

function format_propers(propers: Proper[][], day: Day): ProperDay[] {
  return propers.map((proper) => {
    const [properVernacular, properLatin] = proper;

    const title = properVernacular.title;
    const temporaName = day.getTemporaName();

    const info = {
      id: properVernacular.id,
      title: title || "Feria",
      tempora: temporaName !== title ? temporaName : null,
      rank: properVernacular.rank,
      colors: properVernacular.colors,
      date: day.date,
      day: day,
    };

    return {
      info: info,
      sections: format_proper_section(properLatin, properVernacular),
    };
  });
}

function format_proper_section(
  propers_latin: Proper,
  propers_vernacular: Proper,
) {
  const pl = [];
  const latinPropers = propers_latin.serialize();
  const vernacularPropers = propers_vernacular.serialize();

  for (const vernacular of vernacularPropers) {
    const latinProp = latinPropers.find(
      (latin) => latin?.id === vernacular?.id,
    );
    const latinBody = latinProp?.body;
    const vernacularBody = vernacular?.body;

    pl.push({
      id: vernacular?.id,
      body: {
        latin: latinBody,
        vernacular: vernacularBody,
      },
    });
  }

  return pl;
}

function printAll(calendar: Calendar) {
  return calendar.serialize();
}

export {
  match,
  getCustomPreface,
  getPregeneratedProper,
  yyyyMMDD,
  format_propers,
  printAll,
};
