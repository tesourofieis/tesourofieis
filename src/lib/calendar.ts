import { Day, Observance } from "./day.ts";
import { yyyyMMDD } from "./utils.ts";

import {
  ADVENT,
  EMBER_DAYS_SEPTEMBER,
  FROM_PRE_LENT_TO_POST_PENTECOST,
  LOCAL,
  NATIVITY_OCTAVE_SUNDAY,
  OUTRO,
  POST_EPIPHANY,
  SANCTI,
  SANCTI_10_DUr,
  TEMPORA_NAT2_0,
  WEEK_24_AFTER_PENTECOST,
} from "./constants.ts";
import { rules } from "./rules.ts";

import { UTCDate } from "@date-fns/utc";
import {
  addDays,
  format,
  getDate,
  getDay,
  getMonth,
  getYear,
  isSameDay,
  isSunday,
  nextSunday,
  nextWednesday,
  previousSaturday,
  previousSunday,
  subDays,
} from "date-fns";

class Calendar {
  private container: Map<string, Day>;
  public year: number;

  constructor(year: number) {
    this.year = year;
    this.container = new Map();
    this.buildEmptyCalendar(year);
    this.create();
  }

  private buildEmptyCalendar(year: number): void {
    let date = new UTCDate(year, 0, 1);

    while (date.getFullYear() === year) {
      this.container.set(yyyyMMDD(date), new Day(yyyyMMDD(date)));
      date = addDays(date, 1);
    }
  }

  private create() {
    this.fillInTemporaDays();
    this.fillInSanctiDays();
    this.resolveConcurrency();
  }

  private fillInTemporaDays() {
    // """
    // Days depending on variable date, such as Easter or Advent
    // """
    // # Inserting blocks
    this.insertBlock(this.calcHolyFamily(this.year), POST_EPIPHANY);
    this.insertBlock(
      this.calcSeptuagesima(this.year),
      FROM_PRE_LENT_TO_POST_PENTECOST,
    );
    this.insertBlock(
      this.calcSaturdayBefore24SundayAfterPentecost(this.year),
      POST_EPIPHANY,
      true,
      false,
      this.calcFirstAdventSunday(this.year),
    );
    this.insertBlock(
      this.calc24SundayAfterPentecost(this.year),
      WEEK_24_AFTER_PENTECOST,
    );
    this.insertBlock(
      this.calcFirstAdventSunday(this.year),
      ADVENT,
      false,
      false,
      new UTCDate(this.year, 11, 23),
    );
    this.insertBlock(
      this.calcEmberWednesdaySeptember(this.year),
      EMBER_DAYS_SEPTEMBER,
    );

    // # Inserting single days
    const holyNameDate = this.calcHolyName(this.year);

    this.container.get(yyyyMMDD(holyNameDate)).celebration = [
      new Observance(TEMPORA_NAT2_0, yyyyMMDD(holyNameDate)),
    ];

    const christKingDate = this.calcChristKing(this.year);
    if (this.container.get(yyyyMMDD(christKingDate))) {
      this.container.get(yyyyMMDD(christKingDate)).celebration = [
        new Observance(SANCTI_10_DUr, yyyyMMDD(christKingDate)),
      ];
    }

    const christmasOctaveSunday = this.calcSundayChristmasOctave(this.year);
    if (christmasOctaveSunday) {
      this.insertBlock(christmasOctaveSunday, NATIVITY_OCTAVE_SUNDAY);
    }
  }

  private insertBlock(
    date: Date,
    block: string[],
    reverse = false,
    overwrite = true,
    stopDate?: Date,
  ) {
    if (reverse) {
      // TODO: use toReversed in order not to mutate the original
      // or slice
      block = block.slice().reverse();
    }

    for (const [ii, observanceIds] of block.entries()) {
      const currentDate = addDays(new UTCDate(date), reverse ? -ii : ii);

      if (!observanceIds) {
        continue;
      }

      const dateKey = yyyyMMDD(currentDate);

      if (this.container.get(dateKey)?.celebration.length && !overwrite) {
        break;
      }

      if (isSameDay(stopDate, subDays(currentDate, 1))) {
        break;
      }

      if (this.container.get(dateKey)) {
        this.container.get(dateKey).tempora = [
          new Observance(observanceIds, dateKey),
        ];
        this.container.get(dateKey).celebration = [
          ...this.container.get(dateKey).tempora,
        ];
      }
    }
  }

  private fillInSanctiDays() {
    for (const [date] of this.container.entries()) {
      const dateId = format(date, "MM-dd");
      const days = SANCTI.filter((ii) => ii.startsWith(`santos:${dateId}`)).map(
        (ii) => new Observance(ii, date),
      );

      this.container.get(date)?.celebration.push(...days);

      const local = LOCAL.filter((ii) => ii.startsWith(`santos:${dateId}`)).map(
        (ii) => new Observance(ii, date),
      );

      if (local.length) {
        this.container.get(date)?.local.push(...local);
      }

      const outro = OUTRO.filter((ii) => ii.startsWith(`santos:${dateId}`)).map(
        (ii) => new Observance(ii, date),
      );

      if (outro.length) {
        this.container.get(date)?.outro.push(...outro);
      }
    }
  }

  // Apply `kalendar.rules.*` to the initially instantiated Calendar to fix the situations
  // where more than one Observance falls in the same day.
  private resolveConcurrency() {
    const shiftedAll: { [date: string]: Observance[] } = {};

    for (const [date] of this.container.entries()) {
      const [celebration, commemoration, shifted] = this.applyRules(
        date,
        shiftedAll[date] || [],
      );

      this.container.get(date).celebration = celebration as Observance[];
      this.container.get(date).commemoration = commemoration as Observance[];

      if (shifted) {
        // @ts-ignore
        for (const [k, v] of shifted) {
          if (!shiftedAll[k]) {
            shiftedAll[k] = [];
          }
          shiftedAll[k].push(...(Array.isArray(v) ? v : [v]));
        }
      }
    }
  }

  private applyRules(date: string, shifted: Observance[]) {
    for (const rule of rules) {
      const results = rule(
        this,
        date,
        this.container.get(date)?.tempora,
        this.container.get(date).celebration.concat(shifted),
      );

      if (results) {
        return results;
      }
    }

    return [this.container.get(date).celebration, [], []];
  }

  private calcEasterSunday(year: number): Date {
    const a = year % 19;
    const b = Math.floor(year / 100);
    const c = year % 100;
    const d = Math.floor(b / 4);
    const e = b % 4;
    const f = Math.floor((b + 8) / 25);
    const g = Math.floor((b - f + 1) / 3);
    const h = (19 * a + b - d - g + 15) % 30;
    const i = Math.floor(c / 4);
    const k = c % 4;
    const l = (32 + 2 * e + 2 * i - h - k) % 7;
    const m = Math.floor((a + 11 * h + 22 * l) / 451);
    const month = Math.floor((h + l - 7 * m + 114) / 31);
    const day = ((h + l - 7 * m + 114) % 31) + 1;
    const easter = new UTCDate(year, month - 1, day);

    return easter;
  }

  private calcHolyFamily(year: number): Date {
    // Feast of the Holy Family - First Sunday after Epiphany (06 January).
    const d = new UTCDate(year, 0, 6); // Months are zero-indexed in JavaScript

    return nextSunday(d);
  }

  private calcSeptuagesima(year: number): Date {
    // """ Septuagesima Sunday.
    //
    // Beginning of the pre-Lenten season (Shrovetide).
    // It's 63 days before Ressurection, ninth Sunday before Easter, the third before Ash Wednesday.
    // First day of the Ressurection Sunday - related block.
    // """
    const easter = this.calcEasterSunday(year);
    return subDays(easter, 63);
  }

  private calcFirstAdventSunday(year: number): Date {
    // """
    // First Sunday of Advent - November 27 if it's Sunday, otherwise closest Sunday.
    // """
    const christmasDay = new UTCDate(year, 11, 25);
    const weekday = getDay(christmasDay);

    switch (weekday) {
      case 0: // Sunday
        return new UTCDate(getYear(christmasDay), 10, 27);
      case 1: // Monday
        return new UTCDate(getYear(christmasDay), 11, 3);
      case 2: // Tuesday
        return new UTCDate(getYear(christmasDay), 11, 2);
      case 3: // Wednesday
        return new UTCDate(getYear(christmasDay), 11, 1);
      case 4: // Thursday
        return new UTCDate(getYear(christmasDay), 10, 30);
      case 5: // Friday
        return new UTCDate(getYear(christmasDay), 10, 29);
      default: // Saturday
        return new UTCDate(getYear(christmasDay), 10, 28);
    }
  }

  private calc24SundayAfterPentecost(year: number): Date {
    // """ 24th Sunday after Pentecost.
    //
    // Last Sunday before First Sunday of Advent.
    // It will be always TEMPORA_PENT24_0, which will be placed either:
    // * instead of TEMPORA_PENT23_0
    //   if the number of TEMPORA_PENT*_0 Sundays in given year == 23)
    // * directly after a week starting with TEMPORA_PENT23_0
    //   if the number of TEMPORA_PENT*_0 Sundays in given year == 24)
    // * directly after a week starting with TEMPORA_EPI6_0 (moved from post-epiphania period)
    //   if the number of TEMPORA_PENT*_0 Sundays in given year > 24)
    // """
    return previousSunday(new UTCDate(this.calcFirstAdventSunday(year)));
  }

  private calcSaturdayBefore24SundayAfterPentecost(year: number): Date {
    // """ Last Saturday before 24th Sunday after Pentecost.
    //
    // This is the end of potentially "empty" period that might appear
    // between 23rd and 24th Sunday after Pentecost if Easter is early.
    // In such case one or more Sundays after Epiphany (TEMPORA_EPI*_0) are moved here to "fill the gap"

    return previousSaturday(new UTCDate(this.calc24SundayAfterPentecost(year)));
  }

  private calcEmberWednesdaySeptember(year: number): Date {
    // """ Wednesday of the Ember Days of September.
    //
    // Ember Wednesday in September is a Wednesday after third Sunday
    // of September according to John XXIII's motu proprio
    // "Rubricarum instructum" of June 25 1960.
    // """
    let d = new UTCDate(year, 8, 1);
    while (getMonth(d) === 8) {
      if (isSunday(d) && getDate(d) >= 15 && getDate(d) <= 21) {
        break;
      }
      d = addDays(d, 1);
    }
    return nextWednesday(d);
  }

  private calcHolyName(year: number) {
    // """ The Feast of the Holy Name of Jesus.
    //
    // Kept on the First Sunday of the year; but if this Sunday falls on
    // 1st, 6th or 7th January, the feast is kept on 2nd January.
    // """
    let d = new UTCDate(year, 0, 1);

    while (getDate(d) <= 7) {
      if (isSunday(d)) {
        if (getDate(d) === 1 || getDate(d) === 6 || getDate(d) === 7) {
          return new UTCDate(year, 0, 2);
        }

        return d;
      }

      d = addDays(d, 1);
    }

    console.error("Unexpected condition: Holy Name date not found.");
  }

  private calcChristKing(year: number): Date {
    // The Feast of Christ the King, last Sunday of October.
    const d = new UTCDate(year, 10, 1);

    return previousSunday(d);
  }

  private calcSundayChristmasOctave(year: number): Date | null {
    // """
    // Sunday within the Octave of Christmas, falls between Dec 26 and Dec 31
    // """
    let d = new UTCDate(year, 11, 27); // December 27
    while (d.getFullYear() === year) {
      if (isSunday(d)) {
        return d;
      }
      d = addDays(d, 1);
    }
    return null;
  }

  get(date: string) {
    return this.container.get(date);
  }

  findDay(observanceId: string) {
    for (const [date_, day] of this.container) {
      if (day.all.some((observance) => observance.id === observanceId)) {
        return [date_, day];
      }
    }
  }

  private items() {
    return this.container.entries();
  }

  serialize(): { [date: string]: Day } {
    const serialized = {};
    for (const [date_, day] of this.items()) {
      serialized[date_] = day.serialize();
    }
    return serialized;
  }
}

export { Calendar };
