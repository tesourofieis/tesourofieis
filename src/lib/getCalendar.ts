import { Calendar } from "./src/calendar.ts";

function getCalendar(year: number) {
  const calendar = new Calendar(year);

  return calendar;
}

function getCalendarDay(year: number, date: string) {
  const calendar = new Calendar(year).get(date);

  return calendar;
}

export { getCalendar, getCalendarDay };
