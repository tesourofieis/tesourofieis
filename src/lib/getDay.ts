import { getYear } from "date-fns";
import { Calendar } from "./calendar.ts";
import { format_propers } from "./utils.ts";

function getDay(date: string) {
  const calendar = new Calendar(getYear(date));

  const day = calendar.get(date);

  const proper = day?.getProper(calendar);

  const format = format_propers(proper);

  return format;
}

export { getDay };
