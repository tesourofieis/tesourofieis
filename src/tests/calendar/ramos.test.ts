import { expect, test } from "vitest";
import { getCalendarDay } from "../../lib/getCalendar";

test("Domingo de Ramos", () => {
  const calendar = getCalendarDay("2024-03-24");
  expect(calendar).toMatchObject({
    date: "2024-03-24",
    celebration: [
      {
        colors: ["r", "v"],
        date: "2024-03-24",
        flexibility: "tempora",
        id: "tempora:quad6-0:1:rv",
        name: "quad6-0",
        link: "missal/quaresma/quad6-0",
        priority: 19,
        rank: 1,
        title: "Domingo de Ramos",
        weekday: 0,
      },
    ],
    commemoration: [],
    tempora: [
      {
        colors: ["r", "v"],
        date: "2024-03-24",
        flexibility: "tempora",
        id: "tempora:quad6-0:1:rv",
        name: "quad6-0",
        link: "missal/quaresma/quad6-0",
        priority: 19,
        rank: 1,
        title: "Domingo de Ramos",
        weekday: 0,
      },
    ],
  });
});
