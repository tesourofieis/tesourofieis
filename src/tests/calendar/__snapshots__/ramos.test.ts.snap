// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Domingo de Ramos 1`] = `
Day {
  "celebration": [
    Observance {
      "colors": [
        "r",
        "v",
      ],
      "date": "2024-03-24",
      "flexibility": "tempora",
      "id": "tempora:quad6-0:1:rv",
      "link": "missal/quaresma/quad6-0",
      "name": "quad6-0",
      "priority": 19,
      "rank": 1,
      "title": "Domingo de Ramos",
      "weekday": 0,
    },
  ],
  "commemoration": [],
  "date": "2024-03-24",
  "tempora": [
    Observance {
      "colors": [
        "r",
        "v",
      ],
      "date": "2024-03-24",
      "flexibility": "tempora",
      "id": "tempora:quad6-0:1:rv",
      "link": "missal/quaresma/quad6-0",
      "name": "quad6-0",
      "priority": 19,
      "rank": 1,
      "title": "Domingo de Ramos",
      "weekday": 0,
    },
  ],
}
`;
